staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy.go

```diff
@@ -209,12 +209,55 @@ func (r *responder) Error(_ http.ResponseWriter, _ *http.Request, err error) {
 
 // these methods provide locked access to fields
 
-func (r *proxyHandler) updateAPIService(apiService *apiregistrationapi.APIService) {
+// updateWhileKeepingAvailability updates an api service handling info
+// if this is the first time we see the service we take availability directly from the apiService argument
+// otherwise we take it from the currently stored value
+//
+// note that the current availability of the service is updated through updateAvailabilityIfExists method
+func (r *proxyHandler) updateWhileKeepingAvailability(apiService *apiregistrationapi.APIService) {
+	serviceAvailable := apiregistrationapi.IsAPIServiceConditionTrue(apiService, apiregistrationapi.Available)
+	value := r.handlingInfo.Load()
+	if value != nil {
+		handlingInfo := value.(proxyHandlingInfo)
+		serviceAvailable = handlingInfo.serviceAvailable
+	}
+
+	r.updateAPIService(apiService, serviceAvailable)
+}
+
+// updateAvailabilityIfExists updates availability of a service
+// this method is noop if availability of a service hasn't changed or no service has been registered
+func (r *proxyHandler) updateAvailabilityIfExists(apiService *apiregistrationapi.APIService) {
```

Comment:
trying to wrap my head around the logic here. When can `handlingInfo` be nil? `handlingInfo` needs godocs which answer that. What does it mean not to exist?

Reply:
you are right it cannot be nil, It can be only nil in `updateWhileKeepingAvailability` function when this is the first time we see this service - https://github.com/kubernetes/kubernetes/blob/6e31183d4780ec340af59c333dca6163ac79ad18/staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go#L277